### DISCUSSION TITLE
Why does "go get" give me an old kiali snapshot version

### DISCUSSION THREAD

**Discussion by jshaughn** [kiali/kiali#8015] [General]
I'm creating this discussion on behalf of @zirain , who asked in Slack.  I don't have an immediate answer but wanted to open this up for discussion, to see if we can explain the current behavior and possibly make it predictable.

> do you know why I got [github.com/kiali/kiali](http://github.com/kiali/kiali) v1.55.1-0.20241216195252-fe0584887bd9 when running go get [github.com/kiali/kiali](http://github.com/kiali/kiali)

**Comment by jmazzitelli** [comment]
Just some notes and easy replication procedures.

First, we can confirm that git commit `cb33b859493abf0d2706068ddaaace5656dc0d4d` is the git commit for our 2.2 release; see https://github.com/kiali/kiali/commit/cb33b859493abf0d2706068ddaaace5656dc0d4d

Create a go.mod file in some empty directory and pin kiali to that commit:

```bash
$ mkdir /tmp/test
$ cd /tmp/test
$ cat > go.mod <<EOM
module testgomod

go 1.23.2

require github.com/kiali/kiali cb33b859493abf0d2706068ddaaace5656dc0d4d
EOM
```

Then run go get and notice the odd version it tags it with:

```
$ go get github.com/kiali/kiali
go: downloading github.com/kiali/kiali v1.55.1-0.20241202114854-cb33b859493a
```

Where does that 1.55.1 come from? Notice what `go.mod` now contains:

```
$ grep kiali go.mod
require github.com/kiali/kiali v1.55.1-0.20241202114854-cb33b859493a
```

and what the newly created `go.sum` contains:

```
$ grep kiali go.sum
github.com/kiali/kiali v1.55.1-0.20241202114854-cb33b859493a h1:gpV/Ts2akhN7sUenr/Pba/Hx7K0SgsPpTa6j4qrzws0=
github.com/kiali/kiali v1.55.1-0.20241202114854-cb33b859493a/go.mod h1:y7aFm/xAs0tOR8TDqm59viRIO9t//0zezI1pOH/J9lw=
```

For those wondering, 1.55.1 is git commit `da5430c49f6cee838b2f9f06c4a69f0cfae73f49` as you see here: https://github.com/kiali/kiali/commit/da5430c49f6cee838b2f9f06c4a69f0cfae73f49

**Comment by zirain** [comment]
FYI, can not run go get with following command:
```shell
go get github.com/kiali/kiali@v2.3.0                                     
go: github.com/kiali/kiali@v2.3.0: invalid version: module contains a go.mod file, so module path must match major version ("github.com/kiali/kiali/v2")
```

**Comment by jshaughn** [comment]
I see the same, it seems the v2 releases of Kiali are not available.  And so the commit has used in the repro steps above is not available, and it leads to this 1.55 version.  If I instead use:

```
cat > go.mod <<EOM
module testgomod

go 1.23.2

require (
        github.com/kiali/kiali latest
)
EOM
```

I end up with v1.89.8, which seems to be the most recent release available via the proxy.  I still don't know why 1.55 version shows up as a seeming default, nor do I know why the v2 versions are not available.  Maybe someone actually needs to import .../v2 in their code.

**Comment by jmazzitelli** [comment]
The problem is the Kiali Go code file structure is not conducive to this kind of usage. See:

* https://go.dev/doc/faq#get_version
* https://go.dev/doc/modules/developing#versioning
* https://go.dev/doc/modules/major-version

The Kiali API is not public, and is not meant for public consumption or import. The Kiali Server API is prone to change (often drastically) between versions, and there is no guarantee for backward compatibility at all. This has been true for a while now.

We should probably officially document that, but we have mentioned it in our community forums - here's one time that I know of where we mentioned it: https://github.com/kiali/kiali/discussions/3791#discussioncomment-9883088

This has also been in the code for 2 years now (the main kiali.go file): https://github.com/kiali/kiali/blame/master/kiali.go#L5-L6
```go
// NOTE! The Kiali API is not for public use and is not supported for any use outside of the Kiali UI itself.
// The API can and will change from version to version with no guarantee of backwards compatibility.
```

**Reply by nrfox** [reply]
This also isn't just the API, it's the actual Kiali source code which is even _more_ likely to break between versions. Kiali maintainers don't expect Kiali users to import Kiali as a go library. We expect them to deploy it as an application. When we bumped Kiali to 2.0 it did not align with the go versioning scheme. If Kiali we're a library we would need to have a separate `/v2` folder.

Going back to the original question in this thread, I'm not sure why the psuedo version is `v1.55.1-...` though. It looks weird but if you pin the commit it should still give you the right source.

**Comment by jshaughn** [comment]
Right, the bottom line is that although it's a bit weird to see the `v1.55` version, in general we don't expect people to be using Kiali code like a library.

**Comment by zirain** [comment]
https://github.com/golang/go/issues/71219#issuecomment-2582423822

**Reply by aljesusg** [reply]
Hi sorry to jumo here late..IMO the last time was when the backend and frontend were separated. The fix is easy but ...as others workmates have commented. What is the need/objetive?

The kiali backend is not for public use

**Reply by zirain** [reply]
I understand that your guys don't want pay too much efforts on support this as a lib.

I need to some of existing backend logic(I don't want to reinvent the wheel), the only thing bothering me is an old version cause flaky dependabot alert.

Fine, I can dismiss it.

**Reply by aljesusg** [reply]
I didn't want to say that, it is an easy change that we can make, our concern is more that it is not a library that we are supporting changes since it is for exclusive use for the frontend.

That being said, we will review it

**Reply by nrfox** [reply]
Oh interesting! Thanks for finding out the answer to this @zirain. I wonder if v1.55 tagging master was just an accident since the rest of the tags all reference the release branch.

**Reply by jmazzitelli** [reply]
There was a time we were tagging the master branch (in fact, I think most of the times it was by accident :)) ... our release procedures now no longer tag master branch.

**Reply by aljesusg** [reply]
We can just make a new tag without "v" to let users remove this kind of alerts
